pkg/controller/v1beta1/inferenceservice/reconcilers/hpa/hpa_reconciler.go

```diff
@@ -0,0 +1,173 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hpa
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/kserve/kserve/pkg/apis/serving/v1beta1"
+	"github.com/kserve/kserve/pkg/constants"
+	v2beta2 "k8s.io/api/autoscaling/v2beta2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("HPAReconciler")
+
+//HPAReconciler is the struct of Raw K8S Object
+type HPAReconciler struct {
+	client       client.Client
+	scheme       *runtime.Scheme
+	HPA          *v2beta2.HorizontalPodAutoscaler
+	componentExt *v1beta1.ComponentExtensionSpec
+}
+
+func NewHPAReconciler(client client.Client,
+	scheme *runtime.Scheme,
+	componentMeta metav1.ObjectMeta,
+	componentExt *v1beta1.ComponentExtensionSpec) (*HPAReconciler, error) {
+	hpa, err := createHPA(componentMeta, componentExt)
+	if err != nil {
+		return nil, err
+	}
+	return &HPAReconciler{
+		client:       client,
+		scheme:       scheme,
+		HPA:          hpa,
+		componentExt: componentExt,
+	}, err
+}
+
+func getHPAMetrics(metadata metav1.ObjectMeta) ([]v2beta2.MetricSpec, error) {
+	var metrics []v2beta2.MetricSpec
+	var cpuUtilization int32
+	annotations := metadata.Annotations
+
+	if value, ok := annotations[constants.TargetUtilizationPercentage]; ok {
+		utilization, err := strconv.Atoi(value)
+		if err != nil {
+			return nil, err
+		}
+		cpuUtilization = int32(utilization)
+	} else {
+		cpuUtilization = constants.DefaultCPUUtilization
+	}
+
+	ms := v2beta2.MetricSpec{
+		Type: v2beta2.ResourceMetricSourceType,
+		Resource: &v2beta2.ResourceMetricSource{
+			Name: corev1.ResourceCPU,
+			Target: v2beta2.MetricTarget{
+				Type:               "Utilization",
+				AverageUtilization: &cpuUtilization,
+			},
+		},
+	}
+	metrics = append(metrics, ms)
+	return metrics, nil
+}
+
+func createHPA(componentMeta metav1.ObjectMeta,
+	componentExt *v1beta1.ComponentExtensionSpec) (*v2beta2.HorizontalPodAutoscaler, error) {
+	minReplicas := int32(*componentExt.MinReplicas)
+	if minReplicas < int32(constants.DefaultMinReplicas) {
+		minReplicas = int32(constants.DefaultMinReplicas)
+	}
+	maxReplicas := int32(componentExt.MaxReplicas)
+	if maxReplicas < minReplicas {
+		maxReplicas = minReplicas
+	}
+	metrics, err := getHPAMetrics(componentMeta)
+	if err != nil {
+		return nil, err
+	}
+	hpa := &v2beta2.HorizontalPodAutoscaler{
+		ObjectMeta: componentMeta,
+		Spec: v2beta2.HorizontalPodAutoscalerSpec{
+			ScaleTargetRef: v2beta2.CrossVersionObjectReference{
+				APIVersion: "apps/v1",
+				Kind:       "Deployment",
+				Name:       componentMeta.Name,
+			},
+			MinReplicas: &minReplicas,
```

Comment:
Does HPA support setting minReplica to 0 ?

Reply:
HPA doesn't support scale to 0, let me add checks

Reply:
https://github.com/kubernetes/kubernetes/pull/3247